sbncode/SBNEventWeight/App/SystToolsEventWeight_module.cc

```diff
@@ -0,0 +1,238 @@
+  systtools::ParamHeaderHelper fParamHeaderHelper;
+
+};
+
+
+SystToolsEventWeight::SystToolsEventWeight(fhicl::ParameterSet const& p)
+  : EDProducer{p}
+{
+
+  fGenieModuleLabel = p.get<std::string>("generator_module_label", "generator");
+  fAllowMissingTruth = p.get<bool>("AllowMissingTruth");
+  fDebugMode = p.get<bool>("debugMode");
+
+  fhicl::ParameterSet syst_provider_config = p.get<fhicl::ParameterSet>("generated_systematic_provider_configuration");
+
+  MF_LOG_INFO("SystToolsEventWeight") << "Configuring ISystProvider";
+
+  fSystProviders = systtools::ConfigureISystProvidersFromParameterHeaders(syst_provider_config);
+  fParamHeaderMap = systtools::BuildParameterHeaders(fSystProviders);
+  fParamHeaderHelper = systtools::ParamHeaderHelper(fParamHeaderMap);
+
+  MF_LOG_INFO("SystToolsEventWeight") << "ISystProvider is configuered" << std::endl;
+
+  produces<std::vector<sbn::evwgh::EventWeightMap> >();
+  produces<art::Assns<simb::MCTruth, sbn::evwgh::EventWeightMap> >();
+  produces<std::vector<sbn::evwgh::EventWeightParameterSet>, art::InRun>();
+
+}
+
+
+void SystToolsEventWeight::produce(art::Event& e) {
+
+  auto mcwghvec = std::make_unique<std::vector<sbn::evwgh::EventWeightMap> >();
+  auto wghassns = std::make_unique<art::Assns<simb::MCTruth, sbn::evwgh::EventWeightMap> >();
+
+  art::PtrMaker<sbn::evwgh::EventWeightMap> makeWeightPtr(e);
+
+  std::vector<art::Ptr<simb::MCTruth> > mclist;
+  art::Handle<std::vector<simb::MCTruth>> mcTruthHandle;
+  if(!fGenieModuleLabel.empty()) e.getByLabel(fGenieModuleLabel, mcTruthHandle);
+  // Prooceed even with missing handle if we want to require the MCTruth to be
+  // found, so that an exception will be thrown explaining the problem.
+  if(mcTruthHandle.isValid() || !fAllowMissingTruth){
+
+    art::fill_ptr_vector(mclist, mcTruthHandle);
+
+    if(fDebugMode){
+      std::cout << "[SystToolsEventWeight::produce] mclist.size() = " << mclist.size() << "\n"
+                << "[SystToolsEventWeight::produce] fSystProviders.size() = " << fSystProviders.size() << std::endl;
+    }
+
+    for( auto &sp : fSystProviders ){
+
+      std::unique_ptr<systtools::EventResponse> syst_resp = sp->GetEventResponse(e);
+      if( !syst_resp ){
+        std::cout << "[ERROR]: Got nullptr systtools::EventResponse from provider "
+                  << sp->GetFullyQualifiedName();
+        throw std::exception();
+      }
+
+      //==== syst_resp->size() is (Number of MCTruth)
+      //==== Each index corresponds to each of MCTruth
+      int nMCTruthIndex(0);
+      if(fDebugMode) std::cout << "[SystToolsEventWeight::produce]   syst_resp.size() (= Number of MCTruth) of this SystProvider = " << syst_resp->size() << std::endl;
+
+      //==== Looping over syst_resp is identical to looping over MCTruth
+      for( systtools::EventResponse::iterator itResp = syst_resp->begin(); itResp != syst_resp->end(); ++itResp ){
+
+        systtools::event_unit_response_t resp = *itResp;
```

Comment:
The range-for syntax should work and is better:
```suggestion
      for(systtools::event_unit_response_t const& resp: syst_resp) {
```
In your loop, you effectively do not need the iterator itself. Plus, using `const` makes sure that you don't attempt to change the data by mistake, and the reference avoids a unnecessary copy.

Reply:
Updated as suggested.